src/app-host/app-host.js

```diff
@@ -7,10 +7,61 @@ var telemetry = require('telemetry-helper');
 var cordova;
 var socket = io();
 var nextExecCacheIndex = 0;
+
+// Details of each plugin that has app-host code is injected when this file is served.
+var plugins = {
+    /** PLUGINS **/
+};
+
+var pluginHandlersDefinitions = {
+    /** PLUGIN-HANDLERS **/
+};
+
+var pluginClobberDefinitions = {
+    /** PLUGIN-CLOBBERS **/
+};
+
 var execCache = {};
+var pluginMessages = {};
 var pluginHandlers = {};
+var serviceToPluginMap = {};
+
+function clobber(clobbers, scope, clobberToPluginMap, pluginId) {
+    Object.keys(clobbers).forEach(function (key) {
+        if (clobberToPluginMap && pluginId) {
+            clobberToPluginMap[key] = pluginId;
+        }
+
+        if (clobbers[key] && typeof clobbers[key] === 'object') {
+            scope[key] = scope[key] || {};
+            clobber(clobbers[key], scope[key]);
+        } else {
+            scope[key] = clobbers[key];
+        }
+    });
+}
+
+function applyPlugins(plugins, clobberScope, clobberToPluginMap) {
+    Object.keys(plugins).forEach(function (pluginId) {
+        var plugin = plugins[pluginId];
+        if (plugin) {
+            if (typeof plugin === 'function') {
+                pluginMessages[pluginId] = pluginMessages[pluginId] || new Messages(pluginId, socket);
+                plugin = plugin(pluginMessages[pluginId], exec);
+                plugins[pluginId] = plugin;
+            }
+            if (clobberScope) {
+                clobber(plugin, clobberScope, clobberToPluginMap, pluginId);
+            }
+        }
+    });
+}
 
 function setCordova(originalCordova) {
```

Comment:
Should we rename the function now that it does pretty much all the app-host initialization? The name suggests it only overrides the existing cordova, when in reality it does several extra things (socket handlers for exec calls, telemetry, XHR, touch events and live reload, initializing plugins, etc)... Or, maybe refactor all the non-cordova stuff to a new function? That way we would have a clear separation of cordova initialization and app-host-specific initialization... Something like `initCordova()` and `initAppHost()`?